src/cbor_decode.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+#include "cbor_decode.h"
+
+static uint32_t additional_len(uint8_t additional)
+{
+	switch (additional) {
+	case 24: return 1;
+	case 25: return 2;
+	case 26: return 4;
+	case 27: return 8;
+	default: return 0;
+	}
+}
+
+#define COMP_PTRS(type, p_res, p_min, p_max) \
+		(((p_min == NULL) || (*(type *)p_res >= *(type *)p_min)) \
+		&& ((p_max == NULL) || (*(type *)p_res <= *(type *)p_max)))
+
+static bool value_extract(uint8_t const **const pp_payload,
+		uint8_t const *const p_payload_end, size_t *const p_elem_count,
+		void * const p_result, size_t result_len)
+{
+	uint8_t *p_u8_result  = (uint8_t *)p_result;
+	uint8_t additional = **pp_payload & 0x1F;
+	(*pp_payload)++;
+
+	cbor_decode_trace();
+
+	if (result_len == 0 || *p_elem_count == 0) {
+		cbor_decode_trace();
+		(*pp_payload)--;
+		return false;
+	}
+
+	memset(p_result, 0, result_len);
+#ifdef CONFIG_BIG_ENDIAN
+	p_u8_result[result_len - 1] = additional;
+#else
+	p_u8_result[0] = additional;
+#endif /* CONFIG_BIG_ENDIAN */
+	if (additional > 23) {
+		uint32_t len = additional_len(additional);
+
+		if (len > result_len) {
+			cbor_decode_trace();
+			(*pp_payload)--;
+			return false;
+		}
+#ifdef CONFIG_BIG_ENDIAN
+		memcpy(&p_u8_result[result_len - len], *pp_payload, len);
+#else
+		for (uint32_t i = 0; i < len; i++) {
+			p_u8_result[i] = (*pp_payload)[len - i - 1];
+		}
+#endif /* CONFIG_BIG_ENDIAN */
+		(*pp_payload) += len;
+	}
+
+	if (*pp_payload > p_payload_end) {
+		/* Value extended beyond the payload. */
+		cbor_decode_trace();
+		(*pp_payload)--;
+		return false;
+	}
+	(*p_elem_count)--;
+	return true;
+}
+
+
+static bool int32_decode(uint8_t const **const pp_payload,
+			uint8_t const *const p_payload_end,
+			size_t *const p_elem_count, int32_t *p_result,
+			void *p_min_value, void *p_max_value)
+{
+	uint8_t major_type = **pp_payload >> 5;
+
+	if (!value_extract(pp_payload, p_payload_end, p_elem_count, p_result,
+			4)) {
```

Comment:
Funky wrapping. Maybe one param per line instead?

Reply:
Changed wrapping to wrap earlier.